interfaces/cython/cantera/composite.py

```diff
@@ -654,6 +656,35 @@ def __call__(self, *species):
         return SolutionArray(self._phase[species], states=self._states,
                              extra=self._extra)
 
+    def __len__(self) -> int:
+        return self._shape[0]
+
+    @property
+    def ndim(self) -> int:
+        """The number of dimensions in the SolutionArray.
+
+        .. versionadded:: 3.0
+        """
+        return len(self.shape)
+
+    @property
+    def shape(self) -> tuple[int, ...]:
+        """The shape of the SolutionArray.
+
+        .. versionadded: 3.0
+
+        :return: A tuple of integers with the number of elements in each dimension.
+        """
+        return self._shape
+
+    @property
+    def size(self) -> int:
+        """The number of elements in the SolutionArray.
+
+        .. versionadded: 3.0
+        """
+        return math.prod(self.shape)
```

Comment:
@speth @ischoegl The reason the tests are failing for Python 3.7 is that `math.prod` was added in 3.8, see: https://docs.python.org/3/library/math.html#math.prod

I suppose I'll switch it to NumPy. The reason I chose `math.prod` was that the [NumPy docs](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.size.html) for `ndarray.size` specify that a standard arbitrary size Python `int` is returned from `ndarray.size`, similar to the result from `math.prod`, as opposed to `np.prod` that returns an `np.int_` type. It _seems_ unlikely that this will be a problem, but when trying to match an API, I think it's useful to match it as closely as possible.

The real question here is, is this worth dropping 3.7 support over? Thoughts are welcome :smile: